File: backend/scripts/profile_request_body_buffering.py
#!/usr/bin/env python3
"""Profile Koaryu's request limiter and Starlette parsers with synthetic bodies."""

from __future__ import annotations

import argparse
import asyncio
import gc
import hashlib
import json
import os
import resource
import statistics
import subprocess
import sys
import time
import tracemalloc
from dataclasses import dataclass
from pathlib import Path
from typing import Any

BACKEND_ROOT = Path(__file__).resolve().parents[1]
if str(BACKEND_ROOT) not in sys.path:
    sys.path.insert(0, str(BACKEND_ROOT))

from app.core.request_body_limits import (  # noqa: E402
    CSV_IMPORT_REQUEST_MAX_BYTES,
    DEFAULT_API_REQUEST_MAX_BYTES,
    RequestBodyLimitMiddleware,
)
from app.core.upload_limits import CSV_IMPORT_MAX_BYTES  # noqa: E402
from starlette.exceptions import HTTPException  # noqa: E402
from starlette.formparsers import MultiPartException, MultiPartParser  # noqa: E402
from starlette.requests import Request  # noqa: E402

MIB = 1024 * 1024
ASGI_CHUNK_BYTES = 64 * 1024
BOUNDARY = b"koaryu-profile-boundary"
FILE_PATTERN = b"0123456789abcdef"
TEXT_PATTERN = b"a"


@dataclass(frozen=True)
class Segment:
    pattern: bytes
    length: int


@dataclass(frozen=True)
class Scenario:
    name: str
    kind: str
    segments: tuple[Segment, ...]
    limit_bytes: int
    expected_status: int
    declared_length: int | None
    path: str

    @property
    def body_bytes(self) -> int:
        return sum(segment.length for segment in self.segments)


class SegmentedBody:
    """Lazily materialize a deterministic body in Uvicorn-sized chunks."""

    def __init__(self, segments: tuple[Segment, ...]):
        self._segments = segments
        self._segment_index = 0
        self._segment_offset = 0
        self.remaining = sum(segment.length for segment in segments)
        self.produced = 0
        self.digest = hashlib.sha256()

    def read(self, max_bytes: int) -> bytes:
        if self.remaining == 0:
            return b""

        chunks: list[bytes] = []
        wanted = min(max_bytes, self.remaining)
        while wanted:
            segment = self._segments[self._segment_index]
            available = segment.length - self._segment_offset
            take = min(wanted, available)
            pattern = segment.pattern
            pattern_offset = self._segment_offset % len(pattern)
            repeats = (pattern_offset + take + len(pattern) - 1) // len(pattern)
            materialized = (pattern * repeats)[pattern_offset : pattern_offset + take]
            chunks.append(materialized)
            self._segment_offset += take
            self.remaining -= take
            self.produced += take
            wanted -= take
            if self._segment_offset == segment.length:
                self._segment_index += 1
                self._segment_offset = 0

        body = b"".join(chunks)
        self.digest.update(body)
        return body


def fixed(value: bytes) -> Segment:
    return Segment(value, len(value))


def repeated(pattern: bytes, length: int) -> Segment:
    if not pattern or length < 0:
        raise ValueError("Repeated segments need a pattern and non-negative length.")
    return Segment(pattern, length)


def json_segments(total_bytes: int) -> tuple[Segment, ...]:
    prefix = b'{"payload":"'
    suffix = b'"}'
    filler_bytes = total_bytes - len(prefix) - len(suffix)
    if filler_bytes < 0:
        raise ValueError("JSON body is too small.")
    return (fixed(prefix), repeated(TEXT_PATTERN, filler_bytes), fixed(suffix))


def multipart_segments(*, file_bytes: int, total_bytes: int | None = None) -> tuple[Segment, ...]:
    file_prefix = (
        b"--"
        + BOUNDARY
        + b'\r\nContent-Disposition: form-data; name="file"; filename="students.csv"\r\n'
        + b"Content-Type: text/csv\r\n\r\n"
    )
    field_prefix = (
        b"\r\n--"
        + BOUNDARY
        + b'\r\nContent-Disposition: form-data; name="payload"\r\n'
        + b"Content-Type: application/json\r\n\r\n"
    )
    field_json_prefix = b'{"mapping":"'
    field_json_suffix = b'"}'
    closing = b"\r\n--" + BOUNDARY + b"--\r\n"
    fixed_bytes = (
        len(file_prefix)
        + file_bytes
        + len(field_prefix)
        + len(field_json_prefix)
        + len(field_json_suffix)
        + len(closing)
    )
    target_bytes = total_bytes if total_bytes is not None else fixed_bytes
    filler_bytes = target_bytes - fixed_bytes
    if filler_bytes < 0:
        raise ValueError("Multipart target is smaller than its framing.")
    return (
        fixed(file_prefix),
        repeated(FILE_PATTERN, file_bytes),
        fixed(field_prefix),
        fixed(field_json_prefix),
        repeated(TEXT_PATTERN, filler_bytes),
        fixed(field_json_suffix),
        fixed(closing),
    )


def scenarios(*, quick: bool) -> list[Scenario]:
    values = [
        Scenario(
            name="json-64k",
            kind="json",
            segments=json_segments(64 * 1024),
            limit_bytes=DEFAULT_API_REQUEST_MAX_BYTES,
            expected_status=204,
            declared_length=64 * 1024,
            path="/api/v1/profile",
        ),
        Scenario(
            name="json-512k",
            kind="json",
            segments=json_segments(512 * 1024),
            limit_bytes=DEFAULT_API_REQUEST_MAX_BYTES,
            expected_status=204,
            declared_length=512 * 1024,
            path="/api/v1/profile",
        ),
        Scenario(
            name="json-1m",
            kind="json",
            segments=json_segments(DEFAULT_API_REQUEST_MAX_BYTES),
            limit_bytes=DEFAULT_API_REQUEST_MAX_BYTES,
            expected_status=204,
            declared_length=DEFAULT_API_REQUEST_MAX_BYTES,
            path="/api/v1/profile",
        ),
        Scenario(
            name="multipart-256k",
            kind="multipart",
            segments=multipart_segments(file_bytes=256 * 1024),
            limit_bytes=CSV_IMPORT_REQUEST_MAX_BYTES,
            expected_status=204,
            declared_length=None,
            path="/api/v1/students/import/parse",
        ),
        Scenario(
            name="multipart-2m",
            kind="multipart",
            segments=multipart_segments(file_bytes=2 * MIB),
            limit_bytes=CSV_IMPORT_REQUEST_MAX_BYTES,
            expected_status=204,
            declared_length=None,
            path="/api/v1/students/import/parse",
        ),
        Scenario(
            name="multipart-10m",
            kind="multipart",
            segments=multipart_segments(file_bytes=CSV_IMPORT_MAX_BYTES),
            limit_bytes=CSV_IMPORT_REQUEST_MAX_BYTES,
            expected_status=204,
            declared_length=None,
            path="/api/v1/students/import/parse",
        ),
        Scenario(
            name="multipart-max-envelope",
            kind="multipart",
            segments=multipart_segments(
                file_bytes=CSV_IMPORT_MAX_BYTES,
                total_bytes=CSV_IMPORT_REQUEST_MAX_BYTES,
            ),
            limit_bytes=CSV_IMPORT_REQUEST_MAX_BYTES,
            expected_status=400,
            declared_length=CSV_IMPORT_REQUEST_MAX_BYTES,
            path="/api/v1/students/import/parse",
        ),
        Scenario(
            name="json-declared-overflow",
            kind="json",
            segments=json_segments(64 * 1024),
            limit_bytes=DEFAULT_API_REQUEST_MAX_BYTES,
            expected_status=413,
            declared_length=DEFAULT_API_REQUEST_MAX_BYTES + 1,
            path="/api/v1/profile",
        ),
        Scenario(
            name="json-streamed-overflow",
            kind="json",
            segments=json_segments(DEFAULT_API_REQUEST_MAX_BYTES + 1),
            limit_bytes=DEFAULT_API_REQUEST_MAX_BYTES,
            expected_status=413,
            declared_length=None,
            path="/api/v1/profile",
        ),
        Scenario(
            name="multipart-streamed-overflow",
            kind="multipart",
            segments=multipart_segments(
                file_bytes=CSV_IMPORT_MAX_BYTES,
                total_bytes=CSV_IMPORT_REQUEST_MAX_BYTES + 1,
            ),
            limit_bytes=CSV_IMPORT_REQUEST_MAX_BYTES,
            expected_status=413,
            declared_length=None,
            path="/api/v1/students/import/parse",
        ),
    ]
    if quick:
        return [values[0], values[4], values[7], values[8]]
    return values


def scenario_by_name(name: str) -> Scenario:
    for scenario in scenarios(quick=False):
        if scenario.name == name:
            return scenario
    raise ValueError(f"Unknown scenario: {name}")


def rss_mib() -> float:
    max_rss = resource.getrusage(resource.RUSAGE_SELF).ru_maxrss
    divisor = MIB if sys.platform == "darwin" else 1024
    return max_rss / divisor


async def drain_receive(receive: Any, last_more_body: bool) -> None:
    while last_more_body:
        message = await receive()
        if message["type"] != "http.request":
            return
        last_more_body = bool(message.get("more_body", False))


async def run_one(
    middleware: RequestBodyLimitMiddleware,
    scenario: Scenario,
    start: asyncio.Event,
) -> dict[str, Any]:
    source = SegmentedBody(scenario.segments)
    observed_digest = hashlib.sha256()
    observed_bytes = 0
    downstream_called = False
    upload_rolled: bool | None = None
    parser_status = 204
    response_status: int | None = None

    async def receive() -> dict[str, Any]:
        await asyncio.sleep(0)
        chunk = source.read(ASGI_CHUNK_BYTES)
        return {
            "type": "http.request",
            "body": chunk,
            "more_body": source.remaining > 0,
        }

    async def send(message: dict[str, Any]) -> None:
        nonlocal response_status
        if message["type"] == "http.response.start":
            response_status = int(message["status"])

    async def downstream(
        scope: dict[str, Any],
        replay_receive: Any,
        downstream_send: Any,
    ) -> None:
        nonlocal downstream_called, observed_bytes, parser_status, upload_rolled
        downstream_called = True
        last_more_body = True

        async def observing_receive() -> dict[str, Any]:
            nonlocal observed_bytes, last_more_body
            message = await replay_receive()
            if message["type"] == "http.request":
                chunk = message.get("body", b"")
                observed_digest.update(chunk)
                observed_bytes += len(chunk)
                last_more_body = bool(message.get("more_body", False))
            else:
                last_more_body = False
            return message

        request = Request(scope, observing_receive)
        try:
            if scenario.kind == "json":
                parsed = await request.json()
                if not isinstance(parsed.get("payload"), str):
                    parser_status = 500
            else:
                form = await request.form()
                upload = form["file"]
                upload_rolled = bool(getattr(upload.file, "_rolled", False))
                file_bytes = 0
                while chunk := await upload.read(ASGI_CHUNK_BYTES):
                    file_bytes += len(chunk)
                if file_bytes == 0:
                    parser_status = 500
                await form.close()
        except (HTTPException, MultiPartException):
            parser_status = 400
        finally:
            await drain_receive(observing_receive, last_more_body)

        await downstream_send(
            {"type": "http.response.start", "status": parser_status, "headers": []}
        )
        await downstream_send({"type": "http.response.body", "body": b""})

    scope: dict[str, Any] = {
        "type": "http",
        "asgi": {"version": "3.0", "spec_version": "2.3"},
        "http_version": "1.1",
        "method": "POST",
        "scheme": "http",
        "path": scenario.path,
        "raw_path": scenario.path.encode(),
        "query_string": b"",
        "server": ("127.0.0.1", 8001),
        "client": ("127.0.0.1", 50000),
        "profile_downstream": downstream,
        "headers": [
            (
                b"content-type",
                b"application/json"
                if scenario.kind == "json"
                else b"multipart/form-data; boundary=" + BOUNDARY,
            )
        ],
    }
    if scenario.declared_length is not None:
        scope["headers"].append(
            (b"content-length", str(scenario.declared_length).encode())
        )

    await start.wait()
    started = time.perf_counter()
    await middleware(scope, receive, send)
    latency_ms = (time.perf_counter() - started) * 1000
    integrity_ok = (
        downstream_called
        and source.remaining == 0
        and observed_bytes == scenario.body_bytes
        and observed_digest.digest() == source.digest.digest()
    )
    return {
        "status": response_status,
        "latency_ms": latency_ms,
        "integrity_ok": integrity_ok,
        "downstream_called": downstream_called,
        "source_bytes_consumed": source.produced,
        "upload_rolled": upload_rolled,
    }


async def run_worker(
    scenario: Scenario,
    *,
    concurrency: int,
    trace_allocations: bool,
) -> dict[str, Any]:
    gc.collect()
    baseline_rss_mib = rss_mib()
    allocation_baseline = 0
    if trace_allocations:
        tracemalloc.start(1)
        allocation_baseline = tracemalloc.get_traced_memory()[0]
        tracemalloc.reset_peak()

    start = asyncio.Event()

    async def route_profile_request(
        scope: dict[str, Any],
        receive: Any,
        send: Any,
    ) -> None:
        await scope["profile_downstream"](scope, receive, send)

    middleware = RequestBodyLimitMiddleware(
        route_profile_request,
        api_v1_prefix="/api/v1",
    )
    tasks = [
        asyncio.create_task(run_one(middleware, scenario, start))
        for _ in range(concurrency)
    ]

    await asyncio.sleep(0)
    start.set()
    results = await asyncio.gather(*tasks)

    allocation_peak_mib: float | None = None
    if trace_allocations:
        _, allocation_peak = tracemalloc.get_traced_memory()
        allocation_peak_mib = (allocation_peak - allocation_baseline) / MIB
        tracemalloc.stop()

    peak_rss_mib = rss_mib()
    statuses = [result["status"] for result in results]
    latencies = [result["latency_ms"] for result in results]
    expected_integrity = scenario.expected_status not in {413}
    return {
        "profile": "direct-backend",
        "scenario": scenario.name,
        "kind": scenario.kind,
        "body_bytes": scenario.body_bytes,
        "concurrency": concurrency,
        "expected_status": scenario.expected_status,
        "statuses": statuses,
        "status_ok": all(status == scenario.expected_status for status in statuses),
        "byte_integrity_ok": (
            all(result["integrity_ok"] for result in results)
            if expected_integrity
            else all(not result["downstream_called"] for result in results)
        ),
        "source_bytes_consumed": [result["source_bytes_consumed"] for result in results],
        "upload_rolled": [result["upload_rolled"] for result in results],
        "baseline_rss_mib": round(baseline_rss_mib, 3),
        "peak_rss_mib": round(peak_rss_mib, 3),
        "rss_delta_mib": round(peak_rss_mib - baseline_rss_mib, 3),
        "python_allocation_peak_mib": (
            round(allocation_peak_mib, 3) if allocation_peak_mib is not None else None
        ),
        "latency_p50_ms": round(statistics.median(latencies), 3),
        "latency_max_ms": round(max(latencies), 3),
        "starlette_spool_mib": MultiPartParser.spool_max_size / MIB,
        "starlette_max_part_mib": MultiPartParser.max_part_size / MIB,
    }


def worker_command(
    scenario: Scenario,
    *,
    concurrency: int,
    trace_allocations: bool,
) -> list[str]:
    command = [
        sys.executable,
        str(Path(__file__).resolve()),
        "--worker",
        "--scenario",
        scenario.name,
        "--concurrency",
        str(concurrency),
    ]
    if trace_allocations:
        command.append("--trace-allocations")
    return command


def run_subprocess_worker(
    scenario: Scenario,
    *,
    concurrency: int,
    trace_allocations: bool,
) -> dict[str, Any]:
    completed = subprocess.run(
        worker_command(
            scenario,
            concurrency=concurrency,
            trace_allocations=trace_allocations,
        ),
        check=True,
        cwd=BACKEND_ROOT.parent,
        capture_output=True,
        text=True,
        env={**os.environ, "PYTHONHASHSEED": "0"},
    )
    return json.loads(completed.stdout)


def render_markdown(results: list[dict[str, Any]]) -> str:
    lines = [
        "| path | body MiB | c | peak RSS Δ MiB | Python alloc peak MiB | p50 ms | max ms | status | bytes |",
        "| --- | ---: | ---: | ---: | ---: | ---: | ---: | --- | --- |",
    ]
    for result in results:
        lines.append(
            "| {scenario} | {body:.3f} | {concurrency} | {rss:.3f} | {alloc:.3f} | "
            "{p50:.3f} | {maximum:.3f} | {status} | {integrity} |".format(
                scenario=result["scenario"],
                body=result["body_bytes"] / MIB,
                concurrency=result["concurrency"],
                rss=result["rss_delta_mib"],
                alloc=result["python_allocation_peak_mib"],
                p50=result["latency_p50_ms"],
                maximum=result["latency_max_ms"],
                status="ok" if result["status_ok"] else "FAIL",
                integrity="ok" if result["byte_integrity_ok"] else "FAIL",
            )
        )
    return "\n".join(lines)


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("--quick", action="store_true", help="Run a small smoke profile.")
    parser.add_argument("--json", action="store_true", help="Emit JSON instead of Markdown.")
    parser.add_argument("--worker", action="store_true", help=argparse.SUPPRESS)
    parser.add_argument("--scenario", help=argparse.SUPPRESS)
    parser.add_argument("--concurrency", type=int, default=1, help=argparse.SUPPRESS)
    parser.add_argument("--trace-allocations", action="store_true", help=argparse.SUPPRESS)
    return parser.parse_args()


def main() -> int:
    args = parse_args()
    if args.worker:
        if not args.scenario or args.concurrency < 1:
            raise SystemExit("Worker mode requires a scenario and positive concurrency.")
        result = asyncio.run(
            run_worker(
                scenario_by_name(args.scenario),
                concurrency=args.concurrency,
                trace_allocations=args.trace_allocations,
            )
        )
        print(json.dumps(result, sort_keys=True))
        return 0

    results: list[dict[str, Any]] = []
    concurrency_values = (1,) if args.quick else (1, 2, 4)
    for scenario in scenarios(quick=args.quick):
        if scenario.expected_status == 413:
            scenario_concurrency = (1,)
        elif scenario.name == "multipart-max-envelope" and not args.quick:
            scenario_concurrency = (*concurrency_values, 8)
        else:
            scenario_concurrency = concurrency_values
        for concurrency in scenario_concurrency:
            rss_result = run_subprocess_worker(
                scenario,
                concurrency=concurrency,
                trace_allocations=False,
            )
            allocation_result = run_subprocess_worker(
                scenario,
                concurrency=concurrency,
                trace_allocations=True,
            )
            rss_result["python_allocation_peak_mib"] = allocation_result[
                "python_allocation_peak_mib"
            ]
            results.append(rss_result)

    failed = [
        result
        for result in results
        if not result["status_ok"] or not result["byte_integrity_ok"]
    ]
    if args.json:
        print(json.dumps(results, indent=2, sort_keys=True))
    else:
        print(render_markdown(results))
    return 1 if failed else 0


if __name__ == "__main__":
    raise SystemExit(main())

File: frontend/scripts/profile-request-body-buffering.mjs
#!/usr/bin/env node

import { createHash } from "node:crypto";
import { spawnSync } from "node:child_process";
import { fileURLToPath } from "node:url";
import { setImmediate as delayImmediate } from "node:timers/promises";

import { NextRequest } from "next/server.js";

import { POST } from "../src/app/api/proxy/[...path]/route.ts";
import {
  CSV_IMPORT_MAX_BYTES,
  CSV_IMPORT_PROXY_REQUEST_MAX_BYTES,
  DEFAULT_PROXY_REQUEST_MAX_BYTES,
} from "../src/lib/request-body-limits.ts";

const MIB = 1024 * 1024;
const STREAM_CHUNK_BYTES = 64 * 1024;
const BOUNDARY = new TextEncoder().encode("koaryu-profile-boundary");
const FILE_PATTERN = new TextEncoder().encode("0123456789abcdef");
const TEXT_PATTERN = new TextEncoder().encode("a");
const encoder = new TextEncoder();
const originalFetch = globalThis.fetch;
const originalBackendUrl = process.env.BACKEND_API_URL;

function fixed(value) {
  return { pattern: value, length: value.byteLength };
}

function repeated(pattern, length) {
  if (pattern.byteLength === 0 || length < 0) {
    throw new Error("Repeated segments need a pattern and non-negative length.");
  }
  return { pattern, length };
}

function bytes(value) {
  return encoder.encode(value);
}

export function jsonSegments(totalBytes) {
  const prefix = bytes('{"payload":"');
  const suffix = bytes('"}');
  const fillerBytes = totalBytes - prefix.byteLength - suffix.byteLength;
  if (fillerBytes < 0) {
    throw new Error("JSON body is too small.");
  }
  return [fixed(prefix), repeated(TEXT_PATTERN, fillerBytes), fixed(suffix)];
}

export function multipartSegments({ fileBytes, totalBytes = null }) {
  const boundary = new TextDecoder().decode(BOUNDARY);
  const filePrefix = bytes(
    `--${boundary}\r\n` +
      'Content-Disposition: form-data; name="file"; filename="students.csv"\r\n' +
      "Content-Type: text/csv\r\n\r\n"
  );
  const fieldPrefix = bytes(
    `\r\n--${boundary}\r\n` +
      'Content-Disposition: form-data; name="payload"\r\n' +
      "Content-Type: application/json\r\n\r\n"
  );
  const fieldJsonPrefix = bytes('{"mapping":"');
  const fieldJsonSuffix = bytes('"}');
  const closing = bytes(`\r\n--${boundary}--\r\n`);
  const fixedBytes =
    filePrefix.byteLength +
    fileBytes +
    fieldPrefix.byteLength +
    fieldJsonPrefix.byteLength +
    fieldJsonSuffix.byteLength +
    closing.byteLength;
  const targetBytes = totalBytes ?? fixedBytes;
  const fillerBytes = targetBytes - fixedBytes;
  if (fillerBytes < 0) {
    throw new Error("Multipart target is smaller than its framing.");
  }
  return [
    fixed(filePrefix),
    repeated(FILE_PATTERN, fileBytes),
    fixed(fieldPrefix),
    fixed(fieldJsonPrefix),
    repeated(TEXT_PATTERN, fillerBytes),
    fixed(fieldJsonSuffix),
    fixed(closing),
  ];
}

class SegmentedBody {
  constructor(segments) {
    this.segments = segments;
    this.segmentIndex = 0;
    this.segmentOffset = 0;
    this.remaining = segments.reduce((total, segment) => total + segment.length, 0);
    this.produced = 0;
    this.hash = createHash("sha256");
  }

  read(maxBytes) {
    if (this.remaining === 0) {
      return new Uint8Array();
    }

    const output = new Uint8Array(Math.min(maxBytes, this.remaining));
    let outputOffset = 0;
    while (outputOffset < output.byteLength) {
      const segment = this.segments[this.segmentIndex];
      const available = segment.length - this.segmentOffset;
      const take = Math.min(output.byteLength - outputOffset, available);
      for (let index = 0; index < take; index += 1) {
        output[outputOffset + index] =
          segment.pattern[(this.segmentOffset + index) % segment.pattern.byteLength];
      }
      outputOffset += take;
      this.segmentOffset += take;
      this.remaining -= take;
      this.produced += take;
      if (this.segmentOffset === segment.length) {
        this.segmentIndex += 1;
        this.segmentOffset = 0;
      }
    }
    this.hash.update(output);
    return output;
  }

  digest() {
    return this.hash.copy().digest("hex");
  }
}

function bodyBytes(segments) {
  return segments.reduce((total, segment) => total + segment.length, 0);
}

function scenarios(quick) {
  const values = [
    {
      name: "json-64k",
      kind: "json",
      segments: jsonSegments(64 * 1024),
      expectedStatus: 204,
      declaredLength: 64 * 1024,
      path: ["profile"],
    },
    {
      name: "json-512k",
      kind: "json",
      segments: jsonSegments(512 * 1024),
      expectedStatus: 204,
      declaredLength: 512 * 1024,
      path: ["profile"],
    },
    {
      name: "json-1m",
      kind: "json",
      segments: jsonSegments(DEFAULT_PROXY_REQUEST_MAX_BYTES),
      expectedStatus: 204,
      declaredLength: DEFAULT_PROXY_REQUEST_MAX_BYTES,
      path: ["profile"],
    },
    {
      name: "multipart-256k",
      kind: "multipart",
      segments: multipartSegments({ fileBytes: 256 * 1024 }),
      expectedStatus: 204,
      declaredLength: null,
      path: ["students", "import", "parse"],
    },
    {
      name: "multipart-2m",
      kind: "multipart",
      segments: multipartSegments({ fileBytes: 2 * MIB }),
      expectedStatus: 204,
      declaredLength: null,
      path: ["students", "import", "parse"],
    },
    {
      name: "multipart-10m",
      kind: "multipart",
      segments: multipartSegments({ fileBytes: CSV_IMPORT_MAX_BYTES }),
      expectedStatus: 204,
      declaredLength: null,
      path: ["students", "import", "parse"],
    },
    {
      name: "multipart-max-envelope",
      kind: "multipart",
      segments: multipartSegments({
        fileBytes: CSV_IMPORT_MAX_BYTES,
        totalBytes: CSV_IMPORT_PROXY_REQUEST_MAX_BYTES,
      }),
      expectedStatus: 204,
      declaredLength: CSV_IMPORT_PROXY_REQUEST_MAX_BYTES,
      path: ["students", "import", "parse"],
    },
    {
      name: "json-declared-overflow",
      kind: "json",
      segments: jsonSegments(64 * 1024),
      expectedStatus: 413,
      declaredLength: DEFAULT_PROXY_REQUEST_MAX_BYTES + 1,
      path: ["profile"],
    },
    {
      name: "json-streamed-overflow",
      kind: "json",
      segments: jsonSegments(DEFAULT_PROXY_REQUEST_MAX_BYTES + 1),
      expectedStatus: 413,
      declaredLength: null,
      path: ["profile"],
    },
    {
      name: "multipart-streamed-overflow",
      kind: "multipart",
      segments: multipartSegments({
        fileBytes: CSV_IMPORT_MAX_BYTES,
        totalBytes: CSV_IMPORT_PROXY_REQUEST_MAX_BYTES + 1,
      }),
      expectedStatus: 413,
      declaredLength: null,
      path: ["students", "import", "parse"],
    },
  ];
  return quick ? [values[0], values[4], values[7], values[8]] : values;
}

function scenarioByName(name) {
  const scenario = scenarios(false).find((candidate) => candidate.name === name);
  if (!scenario) {
    throw new Error(`Unknown scenario: ${name}`);
  }
  return scenario;
}

function memorySnapshot() {
  const usage = process.memoryUsage();
  return {
    rss: usage.rss,
    heapUsed: usage.heapUsed,
    external: usage.external,
    arrayBuffers: usage.arrayBuffers,
  };
}

function maxMemory(left, right) {
  return Object.fromEntries(
    Object.keys(left).map((key) => [key, Math.max(left[key], right[key])])
  );
}

function deltaMib(peak, baseline, key) {
  return Math.max(0, peak[key] - baseline[key]) / MIB;
}

async function runWorker(scenario, concurrency) {
  process.env.BACKEND_API_URL = "https://synthetic-backend.invalid/api/v1";
  if (globalThis.gc) {
    globalThis.gc();
  }
  await delayImmediate();
  const baseline = memorySnapshot();
  let peak = baseline;
  const timer = setInterval(() => {
    peak = maxMemory(peak, memorySnapshot());
  }, 1);
  const sources = new Map();
  const forwarded = new Map();
  let upstreamCalls = 0;

  globalThis.fetch = async (_url, init) => {
    upstreamCalls += 1;
    const id = new Headers(init.headers).get("idempotency-key");
    const body = new Uint8Array(init.body);
    forwarded.set(id, createHash("sha256").update(body).digest("hex"));
    peak = maxMemory(peak, memorySnapshot());
    await delayImmediate();
    return new Response(null, { status: 204 });
  };

  let releaseStart;
  const start = new Promise((resolve) => {
    releaseStart = resolve;
  });
  const tasks = Array.from({ length: concurrency }, (_, index) => {
    const id = `profile-${index}`;
    const source = new SegmentedBody(scenario.segments);
    sources.set(id, source);
    const stream = new ReadableStream({
      async pull(controller) {
        await delayImmediate();
        const chunk = source.read(STREAM_CHUNK_BYTES);
        if (chunk.byteLength > 0) {
          controller.enqueue(chunk);
        }
        if (source.remaining === 0) {
          controller.close();
        }
        peak = maxMemory(peak, memorySnapshot());
      },
    });
    const headers = new Headers({
      "content-type":
        scenario.kind === "json"
          ? "application/json"
          : `multipart/form-data; boundary=${new TextDecoder().decode(BOUNDARY)}`,
      "idempotency-key": id,
    });
    if (scenario.declaredLength !== null) {
      headers.set("content-length", String(scenario.declaredLength));
    }
    const request = new NextRequest(
      `https://app.example.test/api/proxy/${scenario.path.join("/")}`,
      {
        method: "POST",
        headers,
        body: stream,
        duplex: "half",
      }
    );

    return (async () => {
      await start;
      const started = performance.now();
      const response = await POST(request, {
        params: Promise.resolve({ path: scenario.path }),
      });
      await response.arrayBuffer();
      return { status: response.status, latencyMs: performance.now() - started, id };
    })();
  });

  releaseStart();
  const results = await Promise.all(tasks);
  peak = maxMemory(peak, memorySnapshot());
  clearInterval(timer);
  const statuses = results.map((result) => result.status);
  const latencies = results.map((result) => result.latencyMs).sort((a, b) => a - b);
  const p50 = latencies[Math.floor((latencies.length - 1) / 2)];
  const expectedUpstreamCalls = scenario.expectedStatus === 413 ? 0 : concurrency;
  const integrityOk =
    upstreamCalls === expectedUpstreamCalls &&
    results.every((result) => {
      const source = sources.get(result.id);
      if (scenario.expectedStatus === 413) {
        return !forwarded.has(result.id);
      }
      return (
        source.remaining === 0 &&
        source.produced === bodyBytes(scenario.segments) &&
        forwarded.get(result.id) === source.digest()
      );
    });

  return {
    profile: "next-proxy",
    scenario: scenario.name,
    kind: scenario.kind,
    body_bytes: bodyBytes(scenario.segments),
    concurrency,
    expected_status: scenario.expectedStatus,
    statuses,
    status_ok: statuses.every((status) => status === scenario.expectedStatus),
    byte_integrity_ok: integrityOk,
    source_bytes_consumed: results.map((result) => sources.get(result.id).produced),
    upstream_calls: upstreamCalls,
    baseline_rss_mib: baseline.rss / MIB,
    peak_rss_mib: peak.rss / MIB,
    rss_delta_mib: deltaMib(peak, baseline, "rss"),
    node_allocation_peak_mib:
      deltaMib(peak, baseline, "heapUsed") + deltaMib(peak, baseline, "external"),
    node_array_buffer_peak_mib: deltaMib(peak, baseline, "arrayBuffers"),
    latency_p50_ms: p50,
    latency_max_ms: latencies.at(-1),
  };
}

function workerCommand(scenario, concurrency) {
  return [
    "--experimental-strip-types",
    "--expose-gc",
    fileURLToPath(import.meta.url),
    "--worker",
    "--scenario",
    scenario.name,
    "--concurrency",
    String(concurrency),
  ];
}

function runSubprocessWorker(scenario, concurrency) {
  const completed = spawnSync(process.execPath, workerCommand(scenario, concurrency), {
    cwd: fileURLToPath(new URL("..", import.meta.url)),
    encoding: "utf8",
    env: { ...process.env, NODE_NO_WARNINGS: "1" },
  });
  if (completed.status !== 0) {
    throw new Error(completed.stderr || `Worker exited ${completed.status}`);
  }
  return JSON.parse(completed.stdout);
}

function renderMarkdown(results) {
  const lines = [
    "| path | body MiB | c | peak RSS Δ MiB | Node alloc peak MiB | array buffers Δ MiB | p50 ms | max ms | status | bytes |",
    "| --- | ---: | ---: | ---: | ---: | ---: | ---: | ---: | --- | --- |",
  ];
  for (const result of results) {
    lines.push(
      `| ${result.scenario} | ${(result.body_bytes / MIB).toFixed(3)} | ` +
        `${result.concurrency} | ${result.rss_delta_mib.toFixed(3)} | ` +
        `${result.node_allocation_peak_mib.toFixed(3)} | ` +
        `${result.node_array_buffer_peak_mib.toFixed(3)} | ` +
        `${result.latency_p50_ms.toFixed(3)} | ${result.latency_max_ms.toFixed(3)} | ` +
        `${result.status_ok ? "ok" : "FAIL"} | ` +
        `${result.byte_integrity_ok ? "ok" : "FAIL"} |`
    );
  }
  return lines.join("\n");
}

function parseArgs(argv) {
  const options = {
    quick: argv.includes("--quick"),
    json: argv.includes("--json"),
    worker: argv.includes("--worker"),
    scenario: null,
    concurrency: 1,
  };
  for (let index = 0; index < argv.length; index += 1) {
    if (argv[index] === "--scenario") {
      options.scenario = argv[index + 1];
    }
    if (argv[index] === "--concurrency") {
      options.concurrency = Number(argv[index + 1]);
    }
  }
  return options;
}

async function main() {
  const options = parseArgs(process.argv.slice(2));
  try {
    if (options.worker) {
      if (!options.scenario || !Number.isInteger(options.concurrency) || options.concurrency < 1) {
        throw new Error("Worker mode requires a scenario and positive concurrency.");
      }
      console.log(JSON.stringify(await runWorker(
        scenarioByName(options.scenario),
        options.concurrency
      )));
      return;
    }

    const results = [];
    const concurrencyValues = options.quick ? [1] : [1, 2, 4];
    for (const scenario of scenarios(options.quick)) {
      let values = concurrencyValues;
      if (scenario.expectedStatus === 413) {
        values = [1];
      } else if (scenario.name === "multipart-max-envelope" && !options.quick) {
        values = [...concurrencyValues, 8];
      }
      for (const concurrency of values) {
        results.push(runSubprocessWorker(scenario, concurrency));
      }
    }
    console.log(options.json ? JSON.stringify(results, null, 2) : renderMarkdown(results));
    if (results.some((result) => !result.status_ok || !result.byte_integrity_ok)) {
      process.exitCode = 1;
    }
  } finally {
    globalThis.fetch = originalFetch;
    if (originalBackendUrl === undefined) {
      delete process.env.BACKEND_API_URL;
    } else {
      process.env.BACKEND_API_URL = originalBackendUrl;
    }
  }
}

if (process.argv[1] === fileURLToPath(import.meta.url)) {
  await main();
}
